File: dashboard/lib/risk-classifier.ts
/**
 * CERT Framework EU AI Act Risk Classifier
 * Pure business logic for classifying AI systems according to EU AI Act
 */

import { RiskInputs, RiskOutputs } from '@/types/wizard';

/**
 * Classify AI system risk level according to EU AI Act
 *
 * Checks prohibited uses (Article 5), high-risk systems (Annex III),
 * limited-risk (transparency requirements), and minimal-risk systems.
 */
export function classifyRisk(inputs: RiskInputs): RiskOutputs {
  // Check prohibited uses first (Article 5)
  if (inputs.biometricIdentification) {
    return {
      classification: 'prohibited',
      triggeredCriteria: ['Real-time biometric identification in public spaces (Article 5.1.d)'],
      complianceRequirements: ['System cannot be deployed under EU AI Act'],
      estimatedComplianceCost: { low: 0, high: 0 },
      estimatedTimeMonths: { low: 0, high: 0 },
      prohibitionReason: 'Real-time biometric identification is prohibited except for specific law enforcement cases'
    };
  }

  if (inputs.socialScoring) {
    return {
      classification: 'prohibited',
      triggeredCriteria: ['Social scoring by public authorities or on their behalf (Article 5.1.c)'],
      complianceRequirements: ['System cannot be deployed under EU AI Act'],
      estimatedComplianceCost: { low: 0, high: 0 },
      estimatedTimeMonths: { low: 0, high: 0 },
      prohibitionReason: 'Social scoring systems that lead to detrimental treatment are prohibited'
    };
  }

  if (inputs.manipulativeTechniques) {
    return {
      classification: 'prohibited',
      triggeredCriteria: ['Manipulative or deceptive techniques causing significant harm (Article 5.1.a)'],
      complianceRequirements: ['System cannot be deployed under EU AI Act'],
      estimatedComplianceCost: { low: 0, high: 0 },
      estimatedTimeMonths: { low: 0, high: 0 },
      prohibitionReason: 'AI systems that manipulate human behavior causing significant harm are prohibited'
    };
  }

  if (inputs.exploitVulnerabilities) {
    return {
      classification: 'prohibited',
      triggeredCriteria: ['Exploits vulnerabilities of specific groups causing significant harm (Article 5.1.b)'],
      complianceRequirements: ['System cannot be deployed under EU AI Act'],
      estimatedComplianceCost: { low: 0, high: 0 },
      estimatedTimeMonths: { low: 0, high: 0 },
      prohibitionReason: 'AI systems exploiting vulnerabilities (age, disability, socio-economic situation) are prohibited'
    };
  }

  // Check high-risk criteria (Annex III)
  const triggeredCriteria: string[] = [];

  if (inputs.criticalInfrastructure) {
    triggeredCriteria.push('Safety component in critical infrastructure - energy, water, gas, heating, transport (Annex III.2)');
  }

  if (inputs.educationAccess) {
    triggeredCriteria.push('Determines access to or admission to educational institutions or vocational training (Annex III.3.a)');
  }

  if (inputs.employmentDecisions) {
    triggeredCriteria.push('Recruitment, evaluation, promotion, termination, or task allocation decisions (Annex III.4.a-b)');
  }

  if (inputs.essentialServicesAccess) {
    triggeredCriteria.push('Evaluates credit worthiness, eligibility for insurance, or access to essential services (Annex III.5.a-b)');
  }

  if (inputs.lawEnforcement) {
    triggeredCriteria.push('Law enforcement applications - risk assessment, polygraph, emotion recognition, deep fakes (Annex III.6)');
  }

  if (inputs.migrationAsylumBorder) {
    triggeredCriteria.push('Migration, asylum, or border control management - risk assessment, verification, complaints (Annex III.7)');
  }

  if (inputs.justiceAdministration) {
    triggeredCriteria.push('Administration of justice - researching and interpreting facts and law (Annex III.8)');
  }

  if (inputs.democraticProcesses) {
    triggeredCriteria.push('Influences democratic processes, electoral behavior, or voting outcomes (Annex III.9)');
  }

  // Classify based on triggered criteria
  if (triggeredCriteria.length > 0) {
    return buildHighRiskOutput(triggeredCriteria, inputs);
  }

  // Check for limited risk (transparency requirements)
  if (inputs.decisionsPerYear > 1000 || inputs.affectedIndividuals > 100) {
    return {
      classification: 'limited-risk',
      triggeredCriteria: ['General-purpose AI system with significant user base'],
      complianceRequirements: [
        'Transparency obligations (Article 52)',
        'Disclose that users are interacting with AI',
        'Provide information about capabilities and limitations',
        'Ensure AI-generated content is machine-detectable'
      ],
      estimatedComplianceCost: { low: 5000, high: 25000 },
      estimatedTimeMonths: { low: 1, high: 3 }
    };
  }

  // Default to minimal risk
  return {
    classification: 'minimal-risk',
    triggeredCriteria: [],
    complianceRequirements: [
      'No mandatory requirements under EU AI Act',
      'Voluntary codes of conduct encouraged (Article 69)',
      'General product safety and liability laws apply'
    ],
    estimatedComplianceCost: { low: 0, high: 0 },
    estimatedTimeMonths: { low: 0, high: 0 }
  };
}

/**
 * Build high-risk output with full compliance requirements
 */
function buildHighRiskOutput(triggeredCriteria: string[], inputs: RiskInputs): RiskOutputs {
  const baseCompliance = [
    'Risk management system throughout lifecycle (Article 9)',
    'Data governance and management quality (Article 10)',
    'Technical documentation per Annex IV (Article 11)',
    'Record-keeping - automatic logs, minimum 6 months retention (Article 12)',
    'Transparency and user information (Article 13)',
    'Human oversight measures - design for human intervention (Article 14)',
    'Accuracy, robustness, and cybersecurity requirements (Article 15)',
    'Quality management system (Article 17)',
    'Conformity assessment - internal control (Annex VI) OR third-party (Annex VII)',
    'Registration in EU database before market placement',
    'Fundamental rights impact assessment if high-risk',
    'CE marking and EU declaration of conformity',
    'Post-market monitoring system (Article 72)',
    'Serious incident reporting - 15 days general, 2 days for fundamental rights violations'
  ];

  // Add specific requirements based on triggered criteria
  if (triggeredCriteria.some(c => c.includes('employment') || c.includes('Recruitment'))) {
    baseCompliance.push('Work council or employee representative consultation requirements');
    baseCompliance.push('Notification to candidates and employees about AI use');
    baseCompliance.push('Right to explanation of automated decisions');
  }

  if (triggeredCriteria.some(c => c.includes('law enforcement'))) {
    baseCompliance.push('Enhanced oversight by data protection authorities');
    baseCompliance.push('Logging requirements for all operations performed');
    baseCompliance.push('Special authorization requirements for certain uses');
  }

  if (triggeredCriteria.some(c => c.includes('credit') || c.includes('insurance'))) {
    baseCompliance.push('Right to human review of automated decisions');
    baseCompliance.push('Explanation of decision logic to affected individuals');
  }

  // Cost estimation based on complexity
  const complexityCriteria = triggeredCriteria.length;
  const volumeFactor = inputs.decisionsPerYear > 100000 ? 2 : inputs.decisionsPerYear > 10000 ? 1.5 : 1;
  const affectedFactor = inputs.affectedIndividuals > 10000 ? 1.5 : 1;

  const baseCostLow = 50000 * complexityCriteria * volumeFactor * affectedFactor;
  const baseCostHigh = 200000 * complexityCriteria * volumeFactor * affectedFactor;

  // Time estimation
  const baseTimeMonths = 6 + (complexityCriteria * 2);

  return {
    classification: 'high-risk',
    triggeredCriteria,
    complianceRequirements: baseCompliance,
    estimatedComplianceCost: {
      low: Math.max(baseCostLow, 75000),
      high: Math.min(baseCostHigh, 1000000)
    },
    estimatedTimeMonths: {
      low: Math.min(baseTimeMonths, 6),
      high: Math.max(baseTimeMonths + 6, 12)
    }
  };
}

/**
 * Get default risk inputs for common scenarios
 */
export function getDefaultRiskInputs(scenario: 'chatbot' | 'hr-screening' | 'credit-scoring'): RiskInputs {
  const defaults: Record<string, RiskInputs> = {
    'chatbot': {
      biometricIdentification: false,
      socialScoring: false,
      manipulativeTechniques: false,
      exploitVulnerabilities: false,
      criticalInfrastructure: false,
      educationAccess: false,
      employmentDecisions: false,
      essentialServicesAccess: false,
      lawEnforcement: false,
      migrationAsylumBorder: false,
      justiceAdministration: false,
      democraticProcesses: false,
      decisionsPerYear: 10000,
      affectedIndividuals: 5000
    },
    'hr-screening': {
      biometricIdentification: false,
      socialScoring: false,
      manipulativeTechniques: false,
      exploitVulnerabilities: false,
      criticalInfrastructure: false,
      educationAccess: false,
      employmentDecisions: true,
      essentialServicesAccess: false,
      lawEnforcement: false,
      migrationAsylumBorder: false,
      justiceAdministration: false,
      democraticProcesses: false,
      decisionsPerYear: 5000,
      affectedIndividuals: 10000
    },
    'credit-scoring': {
      biometricIdentification: false,
      socialScoring: false,
      manipulativeTechniques: false,
      exploitVulnerabilities: false,
      criticalInfrastructure: false,
      educationAccess: false,
      employmentDecisions: false,
      essentialServicesAccess: true,
      lawEnforcement: false,
      migrationAsylumBorder: false,
      justiceAdministration: false,
      democraticProcesses: false,
      decisionsPerYear: 50000,
      affectedIndividuals: 100000
    }
  };

  return defaults[scenario];
}

File: cert/compliance_2025/risk_classifier.py
"""
EU AI Act Risk Classification Engine

Automatic classification of AI systems according to EU AI Act risk tiers
based on Annex III use cases and system characteristics.

Risk Levels:
- Unacceptable: Prohibited practices (social scoring, emotion recognition at work, etc.)
- High-Risk: Systems in Annex III domains requiring full compliance
- Limited: Transparency-only obligations (chatbots, deep fakes, etc.)
- Minimal: No specific obligations (spam filters, AI in games, etc.)
"""

import json
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any


class RiskLevel(Enum):
    """EU AI Act risk classification levels."""

    UNACCEPTABLE = "unacceptable"
    HIGH = "high"
    LIMITED = "limited"
    MINIMAL = "minimal"


class UseCase(Enum):
    """Annex III use case categories for high-risk classification."""

    # Biometrics
    BIOMETRIC_ID_REMOTE = "biometric_identification_remote"
    BIOMETRIC_CATEGORIZATION = "biometric_categorization"
    EMOTION_RECOGNITION = "emotion_recognition"

    # Critical Infrastructure
    INFRASTRUCTURE_SAFETY = "critical_infrastructure_safety"
    INFRASTRUCTURE_MANAGEMENT = "critical_infrastructure_management"

    # Education
    EDUCATION_ACCESS = "educational_access_determination"
    EDUCATION_ASSESSMENT = "educational_assessment"
    EDUCATION_CHEATING = "cheating_detection"

    # Employment
    EMPLOYMENT_RECRUITMENT = "recruitment_selection"
    EMPLOYMENT_DECISIONS = "employment_decisions"
    EMPLOYMENT_MONITORING = "worker_monitoring"

    # Essential Services
    CREDIT_SCORING = "credit_scoring"
    INSURANCE_RISK = "insurance_risk_assessment"
    SOCIAL_SERVICES = "social_services_access"

    # Law Enforcement
    LAW_ENFORCEMENT_PROFILING = "law_enforcement_profiling"
    POLYGRAPH = "polygraph_ai"
    CRIME_PREDICTION = "predictive_policing"

    # Migration
    MIGRATION_RISK = "migration_risk_assessment"
    BORDER_CONTROL = "border_control"
    VISA_PROCESSING = "visa_application_processing"

    # Justice
    JUDICIAL_RESEARCH = "judicial_fact_research"
    SENTENCING = "sentencing_recommendation"

    # Electoral
    ELECTORAL_INFLUENCE = "electoral_influence"

    # General Purpose
    GPAI = "general_purpose_ai"
    GPAI_SYSTEMIC = "gpai_systemic_risk"


@dataclass
class RiskIndicator:
    """An indicator contributing to risk assessment."""

    indicator: str
    present: bool
    weight: float  # Contribution to risk score
    description: str


@dataclass
class RiskAssessment:
    """Complete risk assessment result."""

    system_name: str
    risk_level: RiskLevel
    confidence: float
    use_cases: list[UseCase]
    indicators: list[RiskIndicator]
    requirements: list[str]
    deadline: str
    assessment_date: datetime = field(default_factory=datetime.utcnow)
    notes: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary."""
        return {
            "system_name": self.system_name,
            "risk_level": self.risk_level.value,
            "confidence": self.confidence,
            "use_cases": [uc.value for uc in self.use_cases],
            "indicators": [
                {
                    "indicator": i.indicator,
                    "present": i.present,
                    "weight": i.weight,
                    "description": i.description,
                }
                for i in self.indicators
            ],
            "requirements": self.requirements,
            "deadline": self.deadline,
            "assessment_date": self.assessment_date.isoformat(),
            "notes": self.notes,
        }


# Annex III high-risk use cases mapping
ANNEX_III_CATEGORIES = {
    # Category 1: Biometrics
    "biometrics": {
        "use_cases": [
            UseCase.BIOMETRIC_ID_REMOTE,
            UseCase.BIOMETRIC_CATEGORIZATION,
            UseCase.EMOTION_RECOGNITION,
        ],
        "keywords": [
            "biometric",
            "facial recognition",
            "face detection",
            "fingerprint",
            "iris scan",
            "voice recognition",
            "emotion detection",
            "sentiment analysis",
        ],
        "requirements": [
            "Biometric data protection measures",
            "Human oversight for identification",
            "Accuracy validation across demographic groups",
            "Logging of all biometric processing",
        ],
    },
    # Category 2: Critical Infrastructure
    "critical_infrastructure": {
        "use_cases": [
            UseCase.INFRASTRUCTURE_SAFETY,
            UseCase.INFRASTRUCTURE_MANAGEMENT,
        ],
        "keywords": [
            "power grid",
            "energy",
            "water supply",
            "traffic",
            "transport",
            "digital infrastructure",
            "safety critical",
            "scada",
            "ics",
        ],
        "requirements": [
            "Safety impact assessment",
            "Redundancy and fallback systems",
            "Real-time monitoring capabilities",
            "Incident response procedures",
        ],
    },
    # Category 3: Education
    "education": {
        "use_cases": [
            UseCase.EDUCATION_ACCESS,
            UseCase.EDUCATION_ASSESSMENT,
            UseCase.EDUCATION_CHEATING,
        ],
        "keywords": [
            "student assessment",
            "admission",
            "exam",
            "grading",
            "learning",
            "educational evaluation",
            "proctoring",
            "cheating detection",
        ],
        "requirements": [
            "Fairness validation across student groups",
            "Appeal mechanisms",
            "Explanation of decisions",
            "Human review option",
        ],
    },
    # Category 4: Employment
    "employment": {
        "use_cases": [
            UseCase.EMPLOYMENT_RECRUITMENT,
            UseCase.EMPLOYMENT_DECISIONS,
            UseCase.EMPLOYMENT_MONITORING,
        ],
        "keywords": [
            "recruitment",
            "hiring",
            "cv screening",
            "resume",
            "interview",
            "promotion",
            "termination",
            "performance evaluation",
            "employee monitoring",
        ],
        "requirements": [
            "Non-discrimination validation",
            "Transparency to candidates/employees",
            "Human review of significant decisions",
            "Data minimization",
        ],
    },
    # Category 5: Essential Services
    "essential_services": {
        "use_cases": [
            UseCase.CREDIT_SCORING,
            UseCase.INSURANCE_RISK,
            UseCase.SOCIAL_SERVICES,
        ],
        "keywords": [
            "credit score",
            "loan",
            "mortgage",
            "insurance",
            "benefits",
            "social assistance",
            "welfare",
            "risk assessment",
            "financial",
        ],
        "requirements": [
            "Explainability of decisions",
            "Right to human review",
            "Non-discrimination testing",
            "Regular accuracy audits",
        ],
    },
    # Category 6: Law Enforcement
    "law_enforcement": {
        "use_cases": [
            UseCase.LAW_ENFORCEMENT_PROFILING,
            UseCase.POLYGRAPH,
            UseCase.CRIME_PREDICTION,
        ],
        "keywords": [
            "police",
            "law enforcement",
            "criminal",
            "profiling",
            "predictive policing",
            "risk assessment",
            "recidivism",
            "lie detection",
            "evidence",
        ],
        "requirements": [
            "Strict logging requirements",
            "Human decision-making authority",
            "Regular bias audits",
            "Fundamental rights impact assessment",
        ],
    },
    # Category 7: Migration
    "migration": {
        "use_cases": [
            UseCase.MIGRATION_RISK,
            UseCase.BORDER_CONTROL,
            UseCase.VISA_PROCESSING,
        ],
        "keywords": [
            "migration",
            "asylum",
            "visa",
            "border",
            "immigration",
            "refugee",
            "travel document",
            "entry",
            "deportation",
        ],
        "requirements": [
            "Human oversight mandatory",
            "Appeal mechanisms",
            "Non-discrimination safeguards",
            "Logging of all decisions",
        ],
    },
    # Category 8: Justice
    "justice": {
        "use_cases": [
            UseCase.JUDICIAL_RESEARCH,
            UseCase.SENTENCING,
        ],
        "keywords": [
            "court",
            "judicial",
            "legal",
            "sentencing",
            "verdict",
            "legal research",
            "case law",
            "judge",
            "prosecution",
        ],
        "requirements": [
            "Human judicial authority",
            "Transparency to affected parties",
            "Regular accuracy validation",
            "Fundamental rights compliance",
        ],
    },
}

# Prohibited (unacceptable risk) practices
PROHIBITED_PRACTICES = [
    {
        "name": "Social scoring by public authorities",
        "keywords": ["social score", "social credit", "citizen rating", "behavior score"],
        "description": "AI systems that evaluate or classify natural persons based on their social behavior",
    },
    {
        "name": "Exploitation of vulnerabilities",
        "keywords": [
            "exploit vulnerable",
            "target vulnerable",
            "manipulate children",
            "elderly manipulation",
        ],
        "description": "AI systems that exploit vulnerabilities of specific groups",
    },
    {
        "name": "Subliminal manipulation",
        "keywords": ["subliminal", "subconscious manipulation", "hidden influence"],
        "description": "AI systems using subliminal techniques beyond awareness",
    },
    {
        "name": "Real-time biometric identification in public",
        "keywords": ["real-time facial", "live biometric", "public surveillance biometric"],
        "description": "Real-time remote biometric identification in publicly accessible spaces",
    },
    {
        "name": "Emotion recognition in workplace/education",
        "keywords": ["workplace emotion", "employee emotion", "student emotion monitoring"],
        "description": "Emotion recognition systems in workplace or educational institutions",
    },
    {
        "name": "Biometric categorization (sensitive)",
        "keywords": ["race detection", "religion detection", "sexual orientation detection"],
        "description": "Biometric categorization inferring sensitive attributes",
    },
    {
        "name": "Untargeted facial recognition scraping",
        "keywords": ["facial scraping", "untargeted face collection", "mass facial database"],
        "description": "Creating facial recognition databases through untargeted scraping",
    },
]


class AIActRiskClassifier:
    """
    EU AI Act risk classification engine.

    Automatically classifies AI systems according to EU AI Act risk tiers
    based on system metadata, intended use, and domain characteristics.

    Example:
        classifier = AIActRiskClassifier()

        # Classify a system
        assessment = classifier.classify(
            system_name="HR Recruitment Assistant",
            description="AI system for screening job applications",
            domain="employment",
            keywords=["recruitment", "cv screening", "hiring"]
        )

        print(f"Risk Level: {assessment.risk_level.value}")
        for req in assessment.requirements:
            print(f"  - {req}")
    """

    def __init__(
        self,
        strict_mode: bool = True,
        include_gpai: bool = True,
    ):
        """
        Initialize the classifier.

        Args:
            strict_mode: Use conservative classification (default True)
            include_gpai: Consider GPAI classifications
        """
        self.strict_mode = strict_mode
        self.include_gpai = include_gpai

    def classify(
        self,
        system_name: str,
        description: str = "",
        domain: str = "",
        keywords: list[str] | None = None,
        is_gpai: bool = False,
        compute_flops: float | None = None,  # For GPAI systemic risk
        user_count: int | None = None,  # For GPAI systemic risk
        additional_metadata: dict[str, Any] | None = None,
    ) -> RiskAssessment:
        """
        Classify an AI system's risk level.

        Args:
            system_name: Name of the AI system
            description: System description
            domain: Primary domain (e.g., "employment", "healthcare")
            keywords: Relevant keywords describing functionality
            is_gpai: Whether this is a general-purpose AI model
            compute_flops: Training compute for GPAI (for systemic risk)
            user_count: Number of users for GPAI (for systemic risk)
            additional_metadata: Additional classification metadata

        Returns:
            RiskAssessment with full classification details
        """
        keywords = keywords or []
        all_text = f"{system_name} {description} {' '.join(keywords)}".lower()
        indicators: list[RiskIndicator] = []
        detected_use_cases: list[UseCase] = []
        notes: list[str] = []

        # Check for prohibited practices first
        for practice in PROHIBITED_PRACTICES:
            practice_detected = any(kw in all_text for kw in practice["keywords"])
            indicators.append(
                RiskIndicator(
                    indicator=f"Prohibited: {practice['name']}",
                    present=practice_detected,
                    weight=1.0 if practice_detected else 0.0,
                    description=practice["description"],
                )
            )

            if practice_detected:
                return RiskAssessment(
                    system_name=system_name,
                    risk_level=RiskLevel.UNACCEPTABLE,
                    confidence=0.95 if self.strict_mode else 0.85,
                    use_cases=[],
                    indicators=indicators,
                    requirements=["System use is prohibited under EU AI Act"],
                    deadline="Immediate - Prohibited Practice",
                    notes=[f"Detected prohibited practice: {practice['name']}"],
                )

        # Check Annex III high-risk categories
        requirements = []
        for category_name, category_data in ANNEX_III_CATEGORIES.items():
            category_match = domain.lower() == category_name
            keyword_matches = sum(1 for kw in category_data["keywords"] if kw in all_text)

            is_match = category_match or keyword_matches >= 2

            indicators.append(
                RiskIndicator(
                    indicator=f"Annex III: {category_name}",
                    present=is_match,
                    weight=0.8 if is_match else 0.0,
                    description=f"Category match: {category_match}, Keywords: {keyword_matches}",
                )
            )

            if is_match:
                detected_use_cases.extend(category_data["use_cases"])
                requirements.extend(category_data["requirements"])
                notes.append(f"Matched category: {category_name}")

        # Check GPAI
        if is_gpai and self.include_gpai:
            detected_use_cases.append(UseCase.GPAI)

            # Check for systemic risk (>10^25 FLOPs or other criteria)
            systemic_risk = False
            if compute_flops and compute_flops >= 1e25:
                systemic_risk = True
                notes.append(f"GPAI with systemic risk (compute: {compute_flops:.2e} FLOPs)")

            if user_count and user_count >= 10_000_000:
                systemic_risk = True
                notes.append(f"GPAI with systemic risk (users: {user_count:,})")

            if systemic_risk:
                detected_use_cases.append(UseCase.GPAI_SYSTEMIC)
                requirements.extend(
                    [
                        "Model evaluation and testing",
                        "Systemic risk assessment",
                        "Incident reporting",
                        "Cybersecurity measures",
                    ]
                )

            indicators.append(
                RiskIndicator(
                    indicator="GPAI System",
                    present=True,
                    weight=0.5,
                    description=f"General-purpose AI, systemic risk: {systemic_risk}",
                )
            )

        # Determine final risk level
        if detected_use_cases:
            risk_level = RiskLevel.HIGH
            confidence = 0.9
            deadline = "August 2, 2025 - High-Risk System Compliance"
        elif any(kw in all_text for kw in ["chatbot", "assistant", "conversational"]):
            risk_level = RiskLevel.LIMITED
            confidence = 0.75
            deadline = "August 2, 2025 - Transparency Requirements"
            requirements = [
                "Disclose AI nature to users",
                "Transparency for AI-generated content",
                "Information about capabilities and limitations",
            ]
            notes.append("Detected as interactive AI system (limited risk)")
        else:
            risk_level = RiskLevel.MINIMAL
            confidence = 0.6
            deadline = "No specific deadline - Voluntary best practices"
            requirements = [
                "Consider voluntary code of conduct",
                "Basic documentation recommended",
            ]

        # Remove duplicate requirements
        requirements = list(dict.fromkeys(requirements))

        return RiskAssessment(
            system_name=system_name,
            risk_level=risk_level,
            confidence=confidence,
            use_cases=list(set(detected_use_cases)),
            indicators=indicators,
            requirements=requirements,
            deadline=deadline,
            notes=notes,
        )

    def get_requirements_for_level(
        self,
        risk_level: RiskLevel,
    ) -> list[str]:
        """Get all requirements for a given risk level."""
        if risk_level == RiskLevel.UNACCEPTABLE:
            return ["Use is prohibited - discontinue immediately"]

        if risk_level == RiskLevel.HIGH:
            return [
                "Risk management system (Article 9)",
                "Data governance (Article 10)",
                "Technical documentation (Article 11)",
                "Record-keeping (Article 12)",
                "Transparency to users (Article 13)",
                "Human oversight provisions (Article 14)",
                "Accuracy, robustness, cybersecurity (Article 15)",
                "Quality management system (Article 17)",
                "Conformity assessment",
                "EU declaration of conformity",
                "CE marking",
                "Registration in EU database",
            ]

        if risk_level == RiskLevel.LIMITED:
            return [
                "Inform users of AI interaction",
                "Label AI-generated content",
                "Transparency about capabilities",
            ]

        return [
            "No mandatory requirements",
            "Voluntary codes of conduct encouraged",
        ]

    def get_deadline(self, risk_level: RiskLevel) -> str:
        """Get compliance deadline for risk level."""
        deadlines = {
            RiskLevel.UNACCEPTABLE: "February 2, 2025",
            RiskLevel.HIGH: "August 2, 2025",
            RiskLevel.LIMITED: "August 2, 2025",
            RiskLevel.MINIMAL: "No specific deadline",
        }
        return deadlines.get(risk_level, "Unknown")

    def export_assessment(
        self,
        assessment: RiskAssessment,
        filepath: str,
    ) -> bool:
        """Export assessment to JSON file."""
        try:
            with open(filepath, "w") as f:
                json.dump(assessment.to_dict(), f, indent=2)
            return True
        except Exception:
            return False
